### PR TITLE
Add fmod-function to libPMacc

### PR DESCRIPTION
**Pull request by BeyondEspresso**
Recently I used the `fmod()` function at several occasions (TWTS pulse, upcoming fix for the moving Window). So we have to include this math-function in the templates of libPMacc.

Note: The `fmod` function is entirely distinct from the `modf` function, which is already part of libPMacc.

### CODE REVIEW

File: src/libPMacc/include/algorithms/math/defines/fmod.hpp [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+{
+
+template<typename Type>
+struct Fmod;
+
+/**
+ * Equivalent of the modulus-operator for float types.
+ * Returns the floating-point remainder of x / y (rounded towards zero):
+ * fmod = x - tquot * y
+ * Where tquot is the truncated (i.e., rounded towards zero) result of: x / y.
```
Comment (ax3l) [inline src/libPMacc/include/algorithms/math/defines/fmod.hpp]: That is a great documentation but pls try to avoid copy-pasting word-by-word the internet (copyright!)
  http://www.cplusplus.com/reference/cmath/fmod/

Can you rephrase it and set a reference?

**Comment by BeyondEspresso** [comment]
I rephrased the documentation.

File: src/libPMacc/include/algorithms/math/defines/fmod.hpp [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+template<typename Type>
+struct Fmod;
+
+/**
+ * Equivalent to the modulus-operator for float types
+ * returns the floating-point remainder of x / y.
+ * The functionality corresponds to the C++
+ * math function fmod().
+ * For details, see <a href="http://www.cplusplus.com/reference/cmath/fmod/">
```
Comment (ax3l) [inline src/libPMacc/include/algorithms/math/defines/fmod.hpp]: sry fo he correction, pls just skip the html-foo and just write the link :)

**Comment by BeyondEspresso** [comment]
Alright, I removed it.

File: src/libPMacc/include/algorithms/math/floatMath/fmod.tpp [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+template<>
+struct Fmod<float>
+{
+    typedef float result;
+
+    HDINLINE result operator( )(result x, result y)
+    {
+#if __CUDA_ARCH__
+        return ::fmod( x , y );
```
Comment (psychocoderHPC) [inline src/libPMacc/include/algorithms/math/floatMath/fmod.tpp]: please use `fmodf`

File: src/libPMacc/include/algorithms/math/floatMath/fmod.tpp [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+#else
+        return std::fmod( x , y );
```
Comment (psychocoderHPC) [inline src/libPMacc/include/algorithms/math/floatMath/fmod.tpp]: please remove the space before the comma

File: src/libPMacc/include/algorithms/math/doubleMath/fmod.tpp [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+template<>
+struct Fmod<double>
+{
+    typedef double result;
+
+    HDINLINE result operator( )(result x, result y)
+    {
+#if __CUDA_ARCH__
+        return ::fmod( x , y );
```
Comment (psychocoderHPC) [inline src/libPMacc/include/algorithms/math/doubleMath/fmod.tpp]: please remove the space before the comma

File: src/libPMacc/include/algorithms/math/doubleMath/fmod.tpp [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+#else
+        return std::fmod( x , y );
```
Comment (psychocoderHPC) [inline src/libPMacc/include/algorithms/math/doubleMath/fmod.tpp]: please remove the space before the comma

File: src/libPMacc/include/algorithms/math/floatMath/fmod.tpp (hunk 4 above)
Comment (psychocoderHPC) [inline src/libPMacc/include/algorithms/math/floatMath/fmod.tpp]: please remove the space before the comma

File: src/libPMacc/include/algorithms/math/doubleMath/fmod.tpp [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+{
+namespace math
+{
+
+template<>
+struct Fmod<double>
+{
+    typedef double result;
+
+    HDINLINE result operator( )(result x, result y)
```
Comment (psychocoderHPC) [inline src/libPMacc/include/algorithms/math/doubleMath/fmod.tpp]: Please use the same style of spaces in all your new files.

`(A x, B y)` or `( A x, B y )`

**Comment by BeyondEspresso** [comment]
@psychocoderHPC , I included your suggestions.

File: src/libPMacc/include/algorithms/math/floatMath/fmod.tpp (hunk 3 above)
Comment (ax3l) [inline src/libPMacc/include/algorithms/math/floatMath/fmod.tpp]: @psychocoderHPC @BeyondEspresso 
```
error: namespace "std" has no member "fmodf"
```

Comment (ax3l) [inline src/libPMacc/include/algorithms/math/floatMath/fmod.tpp]: looks like someone is `using std`  ;)

you want http://docs.nvidia.com/cuda/cuda-math-api/group__CUDA__MATH__SINGLE.html

File: src/libPMacc/include/algorithms/math/floatMath/fmod.tpp [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+struct Fmod<float>
+{
+    typedef float result;
+
+    HDINLINE result operator( )(result x, result y)
+    {
+#if __CUDA_ARCH__
+        return ::fmodf(x, y);
+#else
+        return std::fmodf(x, y);
```
Comment (ax3l) [inline src/libPMacc/include/algorithms/math/floatMath/fmod.tpp]: that is still `std::fmod` ;)

File: src/libPMacc/include/algorithms/math/doubleMath/fmod.tpp [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+ *
+ * You should have received a copy of the GNU General Public License
+ * and the GNU Lesser General Public License along with libPMacc.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "types.h"
+#include "math.h"
```
Comment (ax3l) [inline src/libPMacc/include/algorithms/math/doubleMath/fmod.tpp]: not needed, just use `<cmath>` below

File: src/libPMacc/include/algorithms/math/floatMath/fmod.tpp [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+ *
+ * You should have received a copy of the GNU General Public License
+ * and the GNU Lesser General Public License along with libPMacc.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "types.h"
+#include "math.h"
```
Comment (ax3l) [inline src/libPMacc/include/algorithms/math/floatMath/fmod.tpp]: remove and only use c++ include below

**Comment by BeyondEspresso** [comment]
I included your suggestions. Note that the `modf`-implementation, which is already part of the dev-branch has the same problem with including `"math.h"`.

**Comment by ax3l** [comment]
I fixed the other includes in #1342
